[css-counter-styles] Implement complex predefined counter styles

nt1m opened the pull request:
#### 231e2671c75e017d86898fcdb1f3e9c1c88de9ac
<pre>
[css-counter-styles] Implement complex predefined counter styles
<a href="https://bugs.webkit.org/show_bug.cgi?id=256154">https://bugs.webkit.org/show_bug.cgi?id=256154</a>
rdar://103021467

Reviewed by Darin Adler.

Implement the following as complex counter style systems:

simp-chinese-informal
simp-chinese-formal
trad-chinese-informal
trad-chinese-formal
ethiopic-numeric

As a result, authors can now create counter styles that extend the above.

Specification: <a href="https://drafts.csswg.org/css-counter-styles/#complex-predefined-counters">https://drafts.csswg.org/css-counter-styles/#complex-predefined-counters</a>

To implement this, we create UA sheet only values for the system descriptors.
The algorithms are just moved over from RenderListMarker, and re-used in both places as static methods on CSSCounterStyle.

* LayoutTests/TestExpectations:
* Source/WebCore/css/CSSCounterStyle.cpp: Move over code from RenderListMarker
(WebCore::counterForSystemCJK):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemEthiopicNumeric):
(WebCore::CSSCounterStyle::initialRepresentation const):
(WebCore::CSSCounterStyle::isInRange const):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp: Pipe the CSS value for system descriptor
(WebCore::CSSCounterStyleDescriptors::areSymbolsValidForSystem):
(WebCore::CSSCounterStyleDescriptors::systemCSSText const):
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp: Enable the new codepath for complex system
(WebCore::isCounterStyleUnsupportedByUserAgent):
* Source/WebCore/css/CSSCounterStyleRule.cpp: Pipe the CSS value for system descriptor
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSProperties.json: Pass in the context, so the parser knows whether it&apos;s in UA sheet mode
* Source/WebCore/css/CSSValueKeywords.in: Add internal CSS values
* Source/WebCore/css/counterStyles.css: Define the new complex counter styles
(@counter-style simp-chinese-informal):
(@counter-style simp-chinese-formal):
(@counter-style trad-chinese-informal):
(@counter-style trad-chinese-formal):
(@counter-style cjk-ideographic):
(@counter-style ethiopic-numeric):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp: Parse the new internal CSS values
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem): Remove outdated FIXME that is handled by CSSCounterStyleDescriptors::areSymbolsValidForSystem
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/RenderListMarker.cpp: Delete code that was moved to CSSCounterStyle, and re-use the methods there
(WebCore::listMarkerText):
(WebCore::toCJKIdeographic): Deleted.
(WebCore::toEthiopicNumeric): Deleted.

Canonical link: <a href="https://commits.webkit.org/263544@main">https://commits.webkit.org/263544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac85cc4ae19219946c8194e59100b87d505f309d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4997 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5123 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6515 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5090 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4990 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5308 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5096 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/5349 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5083 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5164 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6531 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2652 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4478 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9479 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4515 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/4550 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6166 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4961 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4061 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4451 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1205 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8524 "Built successfully") | | | 
| [  ~~🛠 🧪 unsafe-merge~~](https://ews-build.webkit.org/#/builders/22/builds/563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4815 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSCounterStyle.h:
```diff
@@ -78,6 +78,10 @@ class CSSCounterStyle : public RefCounted<CSSCounterStyle>, public CanMakeWeakPt
     void extendAndResolve(const CSSCounterStyle&);
 
+    enum class Formality : bool { Informal, Formal };
+    static String counterForSystemCJK(int, const UChar[17], Formality);
```
Seems risky to pass this as a C-style array since the count is ignored and not checked at compile time. I suggest a `const std::array&` or `Span` instead.

I also think this argument should be named.

darinadler commented on Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
```diff
@@ -432,6 +437,22 @@ String CSSCounterStyleDescriptors::systemCSSText() const
     case System::Extends:
         builder.append(makeString("extends ", m_extendsName));
         break;
+    // FIXME: Should the following be ASSERT_NOT_REACHED() ?
```
The answer to this is whether the code can be reached!

nt1m replied:
It can be reached in theory, but no one will do it. I returned an empty string, which should do the more correct thing here (`CSSCounterStyleRule::cssText()` will avoid writing a line for `system` in this case)

darinadler commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -1613,6 +1613,11 @@ numeric
 symbolic
 additive
 // fixed
+-internal-simp-chinese-informal
```
Why abbreviate simplified to simp here?

nt1m replied:
I copied from the counter style name, but I guess this name isn't public so I can use the full name

darinadler commented on Source/WebCore/css/CSSValueKeywords.in:
```diff
@@ -1613,6 +1613,11 @@ numeric
 // fixed
+-internal-simp-chinese-informal
+-internal-simp-chinese-formal
+-internal-trad-chinese-informal
```
Why abbreviate traditional to trad here?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -8253,8 +8265,6 @@ RefPtr<CSSValue> consumeCounterStyleSystem(CSSParserTokenRange& range)
     }
 
     if (auto ident = consumeIdent<CSSValueExtends>(range)) {
-        // FIXME: (rdar://103020193) "If a @counter-style uses the extends system, it must not contain a symbols or additive-symbols descriptor, or else the @counter-style rule is invalid." (https://www.w3.org/TR/css-counter-styles-3/#extends-system)
```
I'm glad you can remove this but unclear why.

nt1m replied:
This was already addressed by rdar://103020193, will clarify in changelog

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -182,6 +306,44 @@ String CSSCounterStyle::initialRepresentation(int value) const
     case CSSCounterStyleDescriptors::System::Fixed:
         return counterForSystemFixed(value);
+    case CSSCounterStyleDescriptors::System::SimplifiedChineseInformal:
+        static constexpr UChar simplifiedChineseInformalTable[17] = {
```
Can we share these tables with `RenderListMarker.cpp`? Seems unnecessary to have two sets.

nt1m replied:
Sharing is not trivial, and the goal is to eventually remove the code in `RenderListMarker.cpp` (which is only exercised when CSSCounterStyleAtRulesEnabled is off), so I'd rather focus on that goal instead (which this patch gets us closer to!)

nt1m replied:
Actually it shouldn't be too hard to share the tables, taking care of it.

darinadler reviewed (approved):
Looks great!

nt1m committed:
[css-counter-styles] Implement complex predefined counter styles
https://bugs.webkit.org/show_bug.cgi?id=256154
rdar://103021467

Reviewed by Darin Adler.

Implement the following as complex counter style systems:

simp-chinese-informal
simp-chinese-formal
trad-chinese-informal
trad-chinese-formal
ethiopic-numeric

As a result, authors can now create counter styles that extend the above.

Specification: https://drafts.csswg.org/css-counter-styles/#complex-predefined-counters

To implement this, we create UA sheet only values for the system descriptors.
The algorithms are just moved over from RenderListMarker, and re-used in both places as static methods on CSSCounterStyle.

* LayoutTests/TestExpectations:
* Source/WebCore/css/CSSCounterStyle.cpp: Move over code from RenderListMarker
(WebCore::counterForSystemCJK):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemEthiopicNumeric):
(WebCore::CSSCounterStyle::initialRepresentation const):
(WebCore::CSSCounterStyle::isInRange const):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp: Pipe the CSS value for system descriptor
(WebCore::CSSCounterStyleDescriptors::areSymbolsValidForSystem):
(WebCore::CSSCounterStyleDescriptors::systemCSSText const):
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRegistry.cpp: Enable the new codepath for complex system
(WebCore::isCounterStyleUnsupportedByUserAgent):
* Source/WebCore/css/CSSCounterStyleRule.cpp: Pipe the CSS value for system descriptor
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSProperties.json: Pass in the context, so the parser knows whether it's in UA sheet mode
* Source/WebCore/css/CSSValueKeywords.in: Add internal CSS values
* Source/WebCore/css/counterStyles.css: Define the new complex counter styles
(@counter-style simp-chinese-informal):
(@counter-style simp-chinese-formal):
(@counter-style trad-chinese-informal):
(@counter-style trad-chinese-formal):
(@counter-style cjk-ideographic):
(@counter-style ethiopic-numeric):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp: Parse the new internal CSS values
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem): Remove outdated FIXME that is handled by CSSCounterStyleDescriptors::areSymbolsValidForSystem
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/RenderListMarker.cpp: Delete code that was moved to CSSCounterStyle, and re-use the methods there
(WebCore::listMarkerText):
(WebCore::toCJKIdeographic): Deleted.
(WebCore::toEthiopicNumeric): Deleted.

Canonical link: https://commits.webkit.org/263544@main

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -149,9 +146,9 @@ String CSSCounterStyle::counterForSystemAdditive(unsigned value) const
 
     StringBuilder result;
-    auto appendToResult = [&](String symb, unsigned frequency) {
+    auto appendToResult = [&](String symbol, unsigned frequency) {
```
I think this should take `const String&` to save a bit of reference count churn. Obviously not important to do that in this patch.

nt1m replied:
I've addressed this in https://github.com/WebKit/WebKit/pull/13322

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -166,6 +163,183 @@ String CSSCounterStyle::counterForSystemAdditive(unsigned value) const
+
+String CSSCounterStyle::counterForSystemSimplifiedChineseInformal(int value)
+{
+    static constexpr std::array<UChar, 17> simplifiedChineseInformalTable = {
```
I think if we leave out the `17` and the `=` the compiler will count for us.

nt1m replied:
I get "too few template arguments" when I remove the 17

darinadler commented on Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
```diff
@@ -404,36 +409,33 @@ String CSSCounterStyleDescriptors::systemCSSText() const
-        return builder.toString();
-    }
+    if (m_isExtendedResolved)
+        return makeString("extends ", m_extendsName);
```
We could add the `_s` here to help whoever is working on that project.
